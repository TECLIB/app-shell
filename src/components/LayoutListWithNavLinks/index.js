/*
 *  LICENSE
 *
 *  This file is part of app-shell
 *
 *  app-shell is a subproject of Teclib.
 *
 *  app-shell is free software: you can redistribute it and/or
 *  modify it under the terms of the GNU General Public License
 *  as published by the Free Software Foundation; either version 3
 *  of the License, or (at your option) any later version.
 *
 *  app-shell is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *  ------------------------------------------------------------------------------
 *  @author     Hector Rondon (hrondon@teclib.com)
 *  @copyright  Copyright Â© 2018 Teclib. All rights reserved.
 *  @license    GPLv3 https://www.gnu.org/licenses/gpl-3.0.html
 *  @link       https://github.com/TECLIB/app-shell
 *  @link       https://teclib.github.io/app-shell
 *  @link       https://teclib-edition.com/en
 *  ------------------------------------------------------------------------------
 */

import React, {
  PureComponent,
} from 'react'
import PropTypes from 'prop-types'
import {
  NavLink,
} from 'react-router-dom'
import I18n from 'shared/i18n'
import getMode from 'shared/getMode'
import calc100PercentMinus from 'shared/calc100PercentMinus'
// import {
//   slideTop,
// } from 'shared/animations/index'
import ErrorPage from '../ErrorPage'

/**
 * List of nav links
 * @class ListWithNavLinks
 * @extends PureComponent
 */
class ListWithNavLinks extends PureComponent {
  /**
   * Change render of the menu according to the screen resolution
   * @function styleNav
   * @param {string} mode
   * @param {object} history
   */
  static styleNav(mode, history) {
    return (
      mode === 'small'
        ? history.location.pathname.split('/').length > 3 ? {
          display: 'none',
        } : {
          width: '100%',
        } : {}
    )
  }

  /** @constructor */
  constructor(props) {
    super(props)
    this.state = {
      mode: getMode(),
      itemListPaneWidth: getMode() === 'small' ? '100%' : 320,
      styleNav: ListWithNavLinks.styleNav(getMode(), this.props.history),
    }

    window.addEventListener('resize', this.handleResize)
  }

  /**
   * Run the 'slideTop' animation
   * @function componentDidMount
   */
  componentDidMount() {
    // slideTop(this.nav).play()
  }

  /**
   * Remove 'resize' event listener
   * @function componentWillUnmount
   */
  componentWillUnmount() {
    window.removeEventListener('resize', this.handleResize)
  }

  /**
   * Make sure that the state and props are in sync for when it is required
   * @param {object} nextProps
   * @param {object} prevState
   */
  static getDerivedStateFromProps(nextProps, prevState) {
    return {
      ...prevState,
      styleNav: (
        getMode() === 'small' ? nextProps.history.location.pathname.split('/').length > 3 ? {
          display: 'none',
        } : {
          width: '100%',
        } : {}
      ),
    }
  }

  /**
   * Change state according to the resolution of the screen
   * @function handleResize
   */
  handleResize = () => {
    const nextMode = getMode()
    if (this.state.mode !== nextMode) {
      this.setState({
        mode: nextMode,
        itemListPaneWidth: nextMode === 'small' ? '100%' : 320,
        styleNav: ListWithNavLinks.styleNav(nextMode, this.props.history),
      })
    }
  }

  /**
   * Change render of the child according to the screen resolution
   * @function stylesArticle
   */
  stylesArticle() {
    const validWidth = this.state.itemListPaneWidth === '100%' ? 0 : this.state.itemListPaneWidth
    return ({
      width: calc100PercentMinus(validWidth),
      overflowY: 'auto',
    })
  }

  renderComponent = () => {
    const paths = this.props.history.location.pathname.split('/')

    if (paths[paths.length - 1].indexOf('error') === 0) {
      return <ErrorPage />
    }

    if (this.state.mode === 'small' && !this.state.styleNav.display) {
      return null
    }

    return (
      <div style={this.stylesArticle()}>
        {this.props.children}
      </div>
    )
  }

  /**
   * Render component
   * @function render
   */
  render() {
    return (
      <div className="flex-block flex-block--with-scroll flex-block--with-content-pane">
        <nav
          style={this.state.styleNav}
          className="flex-block__list navlinks"
          ref={(nav) => { this.nav = nav }}
        >
          <ul>
            {this.props.routes.map((route, i) => {
              if (route.path !== '/') {
                return (
                  <li key={`ListWithNavLinks-${i.toString()}`}>
                    <NavLink
                      exact
                      to={`${this.props.rootPath}${route.path}`}
                      activeClassName="--active"
                    >
                      {I18n.t(route.name)}
                    </NavLink>
                  </li>
                )
              }
              return ''
            })}
          </ul>
        </nav>

        { this.renderComponent() }

      </div>
    )
  }
}

ListWithNavLinks.propTypes = {
  routes: PropTypes.array.isRequired,
  rootPath: PropTypes.string.isRequired,
  children: PropTypes.element.isRequired,
  history: PropTypes.object.isRequired,
}

export default ListWithNavLinks
