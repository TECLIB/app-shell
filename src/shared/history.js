import { createBrowserHistory as createHistory } from 'history'

let history = createHistory()

export default history