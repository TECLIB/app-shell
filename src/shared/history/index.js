/** @module history */

/** import dependencies */
import {
  createBrowserHistory as createHistory,
  createMemoryHistory,
} from 'history'

const history = function () {
  if (process.env.NODE_ENV !== 'test') {
    /** Get browser history */
    return createHistory()
  }
  return createMemoryHistory('/')
}

/** Export history object */
export default history()
