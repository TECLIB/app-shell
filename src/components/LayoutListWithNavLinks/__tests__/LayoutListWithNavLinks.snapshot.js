import React from 'react'
import renderer from 'react-test-renderer'
import {
  Router,
} from 'react-router-dom'
import history from 'shared/history'
import LayoutListWithNavLinks from '../index'

describe('LayoutListWithNavLinks', () => {
  beforeAll(() => {
    global.Animation = class Animation {
      play = () => {}
    }
    global.KeyframeEffect = class KeyframeEffect {}
  })

  afterAll(() => {
    global.Animation = undefined
    global.KeyframeEffect = undefined
  })

  test('renders LayoutListWithNavLinks', () => {
    const component = renderer.create(
      <Router history={history}>
        <LayoutListWithNavLinks
          rootPath=""
          history={history}
          routes={[
            {
              path: '/',
              name: 'test 1',
              component: <ul />,
              exact: true,
            },
            {
              path: '/test',
              name: 'test 2',
              component: <ol />,
              exact: true,
            },
          ]}
        >
          <div />
        </LayoutListWithNavLinks>
      </Router>,
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
