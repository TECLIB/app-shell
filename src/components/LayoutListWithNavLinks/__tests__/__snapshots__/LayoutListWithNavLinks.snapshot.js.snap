// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LayoutListWithNavLinks renders LayoutListWithNavLinks 1`] = `
<div
  className="flex-block flex-block--with-scroll flex-block--with-content-pane"
>
  <nav
    className="flex-block__list navlinks"
    style={Object {}}
  >
    <ul>
      
      <li>
        <a
          aria-current={null}
          className={undefined}
          href="/test"
          onClick={[Function]}
          style={undefined}
        >
          test 2
        </a>
      </li>
    </ul>
  </nav>
  <div
    style={
      Object {
        "overflowY": "auto",
        "width": "calc(100% - 320px)",
      }
    }
  >
    <div />
  </div>
</div>
`;
